### DISCUSSION TITLE
Calculation of reaction forces on the loading boundary

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#16945] [Q&A General] [answered]
![Load_disp_comparison](https://user-images.githubusercontent.com/40111479/107560510-58fb4380-6bd5-11eb-94dc-cdbc02c2252b.png)
Dear all,
I wonder if there is any default method (possibly through the aux kernels) to compute the reaction forces in the boundary where the loads (displacement controlled) are applied. I know this can be done via the stresses (stress_xx) in the loading direction when multiplied by the cross-sectional area of the boundary, but can be slightly inaccurate at times. ANSYS does a better comparison where the loads are extracted for the respective boundary from the in-built 'Force-reaction' solution module. Attached please see the comparison. However it showed a huge difference when the conventional path of (stress_xx*area) was followed. Any guidance on the correct method to compute the loads in MOOSE would be highly appreciated.
Kind regards,
Arun
![Load_disp_comparison](https://user-images.githubusercontent.com/40111479/107560545-60bae800-6bd5-11eb-9295-1079a5ca9fcd.png)

**Comment by hugary1995** [comment, marked as answer]
You can save the residuals from the StressDivergence kernels into an AuxVariable, and then sum up the residual on whatever boundary you want to evaluate the reaction force there. If you search for "save_in" in the tensor_mechanics module's regression tests you'll find enough examples.

**Reply by abarun22** [reply]
Dear Gary,
Thanks for the suggestion. This seems to be the approach that would potentially give close results with ANSYS and experiments, however i am stuck with trouble when the volumetric locking correction is included, which screws up the results at large strains. The plasticity model at its usual self requires 'volumetric locking correction' option to be included to give correct results at these strains. The stress divergence calculations (done through ADStressDivergenceTensors) however requires this option to be switched OFF it it were to execute correctly. This however gives wrong results (please see the achieved matching in the attached plot). On the other hand setting 'volumetric locking correction' to TRUE gives the following error.

*** ERROR ***
Volumetric locking correction should be set to false for 1-D problems
 
Being a 3D problem, i could not find reasons for this error and would like to know from you whether there is any possible fix. I tried setting dim=3 in the mesh block with out much success. One more issue i wanted to discuss here is the negative jacobian obtained for certain elements during the course of computation (possibly stress divergence), although it does not interfere with the results. I wonder whether this is something that  needs attention. Also there seems to be 'MooseException' raised during the computation of Aux variable (please see the .log file) which needs a further look-in. 

I guess we almost found the correct set of input commands (see the attached .i file) to be included that calculate precisely the reaction forces. If this can be done with locking correction ON the problem is solved. I welcome your suggestions in this regard.
Kind regards,
Arun
[16022021.zip](https://github.com/idaholab/moose/files/5990410/16022021.zip)

**Reply by hugary1995** [reply]
Sorry Arun I'm late on this. Setting `volumetric_locking_correction = true` is the correct thing to do here. The relevant logic for that error can be found in `ADStressDivergenceTensors.C`:
```
if (_ndisp == 1 && _volumetric_locking_correction)
    mooseError("Volumetric locking correction should be set to false for 1-D problems.");
```
So the kernel must think you only coupled one displacement to it. Indeed, in the input file you had `displacements = disp_x` in ADStressDivergenceTensors, therefore it overrides your GlobalParams definition of `displacements = 'disp_x disp_y disp_z'`.

Simply removing `displacements = disp_x` and similarly the lines `displacements = disp_y` and `displacements = disp_z` should work.

**Comment by abarun22** [comment]
Dear all,
I peformed an exercise in setting the elements to SECOND order to bye-pass the problems associated with 'volumetric_locking_correction' so that i get same response as physical tests.  This time i received a seg fault error (as given below), which i wanted to report. Please could you comment if the second order elements are a better way to avoid locking behaviour and achieve correct results.
Also let me know how this memory crash can be reported.

(base) abalasub@freia013> ../../../hardening_model_app-opt -i PieceWise_test.i


Framework Information:
MOOSE Version:           git commit 4c15601 on 2021-02-12
LibMesh Version:         c70ab40900490f44391f1b891c8c3731c45f0cf1
PETSc Version:           3.14.2
SLEPc Version:           3.14.0
Current Time:            Thu Feb 18 11:00:26 2021
Executable Timestamp:    Mon Feb 15 19:33:31 2021

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:
    Total:                 12369
    Local:                 12369
  Elems:
    Total:                 1242
    Local:                 1242
  Num Subdomains:          1
  Num Partitions:          1

Nonlinear System:
  Num DOFs:                37107
  Num Local DOFs:          37107
  Variables:               { "disp_x" "disp_y" "disp_z" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "SECOND"

Auxiliary System:
  Num DOFs:                86787
  Num Local DOFs:          86787
  Variables:               { "stress_xx" "strain_xx_av" "plastic_strain_xx" "vonmises" } { "resid_x"
                             "resid_y" "resid_z" }
  Finite Element Types:    "MONOMIAL" "LAGRANGE"
  Approximation Orders:    "SECOND" "SECOND"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             Preconditioned JFNK
  MOOSE Preconditioner:    SMP

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.

Projecting initial condition                                                               [  1.39 s]
Computing initial stateful property values                                                 [  1.19 s]

Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | VonMises       | disp_xx        | p_xx           | resid_x        | s_xx           | strain_xx_pp   |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+


Time Step 1, time = 0.1, dt = 0.1
Computing initial residual .                                                               [  2.37 s]
 0 Nonlinear |R| = 1.192609e+09
Segmentation fault


Kind regards,
Arun
[PieceWise_test.txt](https://github.com/idaholab/moose/files/6002421/PieceWise_test.txt)

**Comment by abarun22** [comment]
Dear Gary,
We nearly identified the methods (especially the usage of 'ADStressDivergenceTensors') needed to achieve the correct results.  Definition of volumetric locking correction and displacements variables are done in the global parameters block and never allowed to get over-written thereafter. The only thing i wish to highlight was regarding the accuracy of the results computed by this kernel.  Setting OFF 'use_displaced_mesh' produced exact results (as ANSYS and experiments) up to the UTS and showed a marked difference thereafter.  On the other hand setting this option to TRUE gives acceptable close results (as in the attached plot), but not the exact fit one would look for. I still believe there is further scope to improve the accuracy of results, which i am unaware of.  Request you please to take a look at the attached input commands and let me know your suggestions.
![Load_displacement_compare](https://user-images.githubusercontent.com/40111479/109005664-1c821a00-76a2-11eb-9072-c494c89191df.png)
[PieceWise_test.txt](https://github.com/idaholab/moose/files/6036165/PieceWise_test.txt)

Kind regards,
Arun

**Reply by hugary1995** [reply]
Hello Arun,

There are two separate sets of things you are trying to do here:
1. setting up the correct input file, and
2. tuning parameters to match the experiment.

I don't think you can do both at the same time. I would first make sure 1 is done, then 2.
I see that in the input file you are using the tensor mechanics action. That action will set up the correct kernels and strain calculators for you -- so you don't need to define them. In fact, I would just remove the kernel you had in there.

Just so we are on the same page, the input file after modification should look like this:
```
# A UserObject IsotropicSD test, with power rule hardening with rate 1e2.
# Linear strain is applied in the x and y direction.

[Mesh]
  type = FileMesh
  file = Specimen_QtrModel.e
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = FINITE
    incremental = true
    add_variables = true 
    use_automatic_differentiation = true
    save_in = 'resid_x resid_y resid_z'
  [../]
[]

[BCs]
  [./Load_RE]
    type = FunctionDirichletBC
    variable = disp_x
    function = '0.0226*t*(0.15/0.125)'
    boundary = "Right_end"
  [../]
  [./Symm_LE]
    type = ADDirichletBC
    variable = disp_x
    value = 0
    boundary = "Left_end"
  [../]
  [./Symm_BY]
    type = ADDirichletBC
    variable = disp_y
    value = 0.0
    boundary = "Bottom"
  [../]
  [./Symm_LZ]
    type = ADDirichletBC
    variable = disp_z
    value = 0.0
    boundary = "Lateral_face"
  [../]
  [./Const_RY]
    type = ADDirichletBC
    variable = disp_y
    value = 0.0
    boundary = "Right_end"
  [../]
  [./Const_RZ]
    type = ADDirichletBC
    variable = disp_z
    value = 0.0
    boundary = "Right_end"
  [../]
[]

[AuxVariables]
  [./stress_xx]
    order = FIRST
    family = MONOMIAL
  [../]
  [./strain_xx_av]
   order = FIRST
    family = MONOMIAL
  [../]
  [./plastic_strain_xx]
    order = FIRST
    family = MONOMIAL
  [../]
  [./vonmises]
    order = FIRST
    family = MONOMIAL
  [../]
  [./resid_x]
    order = FIRST
    family = LAGRANGE
  [../]
  [./resid_y]
  [../]
  [./resid_z]
  [../]
[] 

[AuxKernels]
  [./stress_xx]
    type = ADRankTwoAux
    rank_two_tensor = stress
    variable = stress_xx
    index_i = 0
    index_j = 0
  [../]
  [./strain_xx_ak]
    type = ADRankTwoAux
    rank_two_tensor = total_strain
    variable = strain_xx_av
    index_i = 0
    index_j = 0
  [../]
  [./vonmises]
    type = ADRankTwoScalarAux
    rank_two_tensor = stress
    variable = vonmises
    scalar_type = VonMisesStress
  [../]
  [./plastic_strain_xx]
    type = ADRankTwoAux
    rank_two_tensor = plastic_strain
    variable = plastic_strain_xx
    index_i = 0
    index_j = 0
  [../]
[] 

[Postprocessors]
  [./disp_xx]
    type = SideAverageValue
    variable = disp_x
    boundary = "Right_end"
  [../]
  [./s_xx]
    type = SideAverageValue
    variable = stress_xx
    boundary = "Right_end"
  [../]
  [./p_xx]
    type = SideAverageValue
    variable = plastic_strain_xx
    boundary = "Left_end"
  [../]
  [./strain_xx_pp]
    type = SideAverageValue
    variable = strain_xx_av
    boundary = "Left_end"
  [../]
  [./VonMises]
    type = SideAverageValue
    variable = vonmises
    boundary = "Right_end"
  [../]
  [./resid_x]
    type = SideAverageValue
    variable = resid_x
    boundary = "Right_end"
  [../]
[]

[Functions]
  [./hf]
    type = PiecewiseLinear
    data_file = PW_last_iter_AsIs_extractedYield_start.csv
    scale_factor = 1
    format = columns
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 2.02E+11
    poissons_ratio = 0.315
  [../]
  [./isotropic_plasticity]
    type =ADIsotropicPlasticityStressUpdate
    yield_stress = 554146341.5
    hardening_function = hf
    use_displaced_mesh=true
  [../]
  [./radial_return_stress]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'isotropic_plasticity'
  [../]
[]

[Executioner]
  num_steps = 500
  dt = 0.1
  type = Transient

  nl_rel_tol = 1e-6
  nl_max_its = 20
  l_tol = 1e-4
  l_max_its = 100

  solve_type = PJFNK
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
[]

[Outputs]
  perf_graph = false
  exodus = true
  csv = true
[]

[Preconditioning]
 [./smp]
   type = SMP
   full = true
 [../]
[]
```
The parameters like "use_displaced_mesh" should be set to give you the correct solution, instead of tuned to give you results that are closer to the experimental data. In the case of finite deformation, we calculate cauchy stress in the stress calculator, so the stress divergence kernels should run on displaced mesh, i.e. use_displaced_mesh = true. Again, you don't need to worry about all this if you are using the action.

**Comment by abarun22** [comment]
Dear Gary,
Thanks for the response. I agree that we must try and set up the correct solution with the input commands rather than matching with the experimental results. But the issue here is that we have variants of solution for different set of commands and the one that produces results close to the experiments is elusive here. If we manage to match the curves very closely, then we can say that we've found the most realistic solution. I would still like to retain the added kernel for stress divergence calculations as it gives a much improved fit as compared to running without them (Please see the attached plot). The fact that we achieve an exact fit without the usage of 'use_displaced_mesh' in the divergence kernel (other plot) atleast for the small strain limit suggests that some other options could play a part here and there is very much hope that we get an exact solution.
Kind regards,
Arun
![Load_displacement_compare](https://user-images.githubusercontent.com/40111479/110360910-3d018b00-8037-11eb-93e9-c566b07694f6.png)
![Load_displacement_WO_defMesh](https://user-images.githubusercontent.com/40111479/110360913-3e32b800-8037-11eb-81c1-1536f7bf6146.png)
